### PR TITLE
Merge inheritance margin feature

### PR DESCRIPTION
**Pull request by Cosifne**

### CODE REVIEW

**Comment by Cosifne** [comment]
: ) looks like there are conflicts. I'll first create a resolving PR then update this

**Comment by Cosifne** [comment]
This includes the inheritance margin feature and needs M2 approval.
@vatsalyaagrawal

**Comment by Cosifne** [comment]
@sharwell @CyrusNajmabadi @ryzngard 
I need one approval to check this in, could anyone help me?

**Comment by sharwell** [comment]
I retargeted to vs-deps due to dependency changes.

File: src/EditorFeatures/CSharp/InheritanceMargin/CSharpInheritanceMarginService.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                // 1. Add the type declaration node.(e.g. class, struct etc..)
+                // Use its identifier's position as the line number, since we want the margin to be placed with the identifier
+                builder.Add(typeDeclarationNode);
+
+                // 2. Add type members inside this type declaration.
+                foreach (var member in typeDeclarationNode.Members)
+                {
+                    if (member.IsKind(
+                        SyntaxKind.MethodDeclaration,
+                        SyntaxKind.PropertyDeclaration,
```
Comment (sharwell) [inline src/EditorFeatures/CSharp/InheritanceMargin/CSharpInheritanceMarginService.cs:44]: ❔ Missing `IndexerDeclaration`?

File: src/EditorFeatures/Core.Wpf/InheritanceMargin/MarginGlyph/InheritanceMarginViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        public static InheritanceMarginViewModel Create(
+            ClassificationTypeMap classificationTypeMap,
+            IClassificationFormatMap classificationFormatMap,
+            InheritanceMarginTag tag)
+        {
+            var members = tag.MembersOnLine;
+            if (members.Length == 1)
+            {
+                var member = tag.MembersOnLine[0];
+                var textAppended = " " + EditorFeaturesWpfResources.is_inherited;
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/InheritanceMargin/MarginGlyph/InheritanceMarginViewModel.cs:63]: ⚠️ This makes assumptions about the grammar structure in the language. Better to have a dedicated resource with placeholders.

File: src/EditorFeatures/Core.Wpf/EditorFeaturesWpfResources.resx [hunk 3]
```diff
@@ -180,4 +180,19 @@
...
+  </data>
+  <data name="Navigate_to_0" xml:space="preserve">
+    <value>Navigate to '{0}'</value>
+  </data>
+  <data name="_0_in_1" xml:space="preserve">
+    <value>'{0}' in '{1}'</value>
+  </data>
+  <data name="Multiple_members_are_inherited_on_line_0" xml:space="preserve">
+    <value>Multiple members are inherited on line {0}</value>
+  </data>
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/EditorFeaturesWpfResources.resx:197]: 💡 Would be good to have a comment that this resource is only used for UI Automation / accessibility

File: src/EditorFeatures/Core.Wpf/InheritanceMargin/InheritanceGlyphFactoryProvider.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            IThreadingContext threadingContext,
+            IStreamingFindUsagesPresenter streamingFindUsagesPresenter,
+            ClassificationTypeMap classificationTypeMap,
+            IClassificationFormatMapService classificationFormatMapService,
+            IWaitIndicator waitIndicator)
+        {
+            _threadingContext = threadingContext;
+            _streamingFindUsagesPresenter = streamingFindUsagesPresenter;
+            _classificationTypeMap = classificationTypeMap;
+            _classificationFormatMap = classificationFormatMapService.GetClassificationFormatMap("tooltip");
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/InheritanceMargin/InheritanceGlyphFactoryProvider.cs:44]: 💡 I'd prefer this be lazily evaluated in `GetGlyphFactory`

File: src/EditorFeatures/Core.Wpf/InheritanceMargin/MarginGlyph/InheritanceMargin.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                        _threadingContext,
+                        _streamingFindUsagesPresenter,
+                        context.CancellationToken));
+            }
+        }
+
+        private void ChangeBorderToHoveringColor()
+        {
+            SetResourceReference(BackgroundProperty, "VsBrush.CommandBarMenuBackgroundGradient");
+            SetResourceReference(BorderBrushProperty, "VsBrush.CommandBarMenuBorder");
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/InheritanceMargin/MarginGlyph/InheritanceMargin.xaml.cs:86]:
```suggestion
            SetResourceReference(BackgroundProperty, VsBrushes.CommandBarMenuBackgroundGradientKey);
            SetResourceReference(BorderBrushProperty, VsBrushes.CommandBarMenuBorderKey);
```

File: src/EditorFeatures/Core.Wpf/InheritanceMargin/MarginGlyph/InheritanceMarginViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                // Same automation name can't be set for control. So add the line number info.
+                var automationName = string.Format(EditorFeaturesWpfResources.Multiple_members_are_inherited_on_line_0, tag.LineNumber);
+                var menuItemViewModels = tag.MembersOnLine
+                    .SelectAsArray(MemberMenuItemViewModel.Create)
+                    .CastArray<InheritanceContextMenuItemViewModel>();
+                return new InheritanceMarginViewModel(tag.Moniker, textBlock, automationName, menuItemViewModels, true);
+            }
+        }
+
+        private static void WrapMemberWithinApostrophe(IList<Inline> memberInlines)
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/InheritanceMargin/MarginGlyph/InheritanceMarginViewModel.cs:90]: ⚠️ This makes assumptions about the language. Some translations prefer `"` to `'`.

File: src/EditorFeatures/Core.Wpf/Microsoft.CodeAnalysis.EditorFeatures.Wpf.csproj [hunk 7]
```diff
@@ -37,6 +37,7 @@
     <PackageReference Include="Microsoft.VisualStudio.Text.UI.Wpf" Version="$(MicrosoftVisualStudioTextUIWpfVersion)" />
     <PackageReference Include="Microsoft.VisualStudio.SDK.EmbedInteropTypes" Version="$(MicrosoftVisualStudioSDKEmbedInteropTypesVersion)" />
     <PackageReference Include="Microsoft.VisualStudio.Threading" Version="$(MicrosoftVisualStudioThreadingVersion)" />
+    <PackageReference Include="Microsoft.VisualStudio.Shell.15.0" Version="$(MicrosoftVisualStudioShell150Version)" />
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/Microsoft.CodeAnalysis.EditorFeatures.Wpf.csproj:40]: ❔ Shouldn't code depending on this should move up to the VS layer

File: src/EditorFeatures/Core/FindUsages/IDefinitionsAndReferencesFactory.cs [hunk 8]
```diff
@@ -69,6 +69,17 @@ internal static class DefinitionItemExtensions
...
 
+        public static Task<DefinitionItem> ToNonClassifiedDefinitionItemAsync(
+            this ISymbol definition,
+            Solution solution,
+            bool includeHiddenLocations,
+            CancellationToken cancellationToken)
+        {
+            return ToDefinitionItemAsync(
+                definition, solution, isPrimary: false, includeHiddenLocations, includeClassifiedSpans: false,
+                options: FindReferencesSearchOptions.Default, cancellationToken: cancellationToken);
```
Comment (sharwell) [inline src/EditorFeatures/Core/FindUsages/IDefinitionsAndReferencesFactory.cs:80]: ❔ What does `FindReferencesSearchOptions` get used for here? Is there any reason why we use `FindReferencesSearchOptions.Default` without setting `UnidirectionalHierarchyCascade` (the latter performs better)

File: src/EditorFeatures/Core/GoToDefinition/GoToDefinitionHelpers.cs [hunk 9]
```diff
@@ -129,5 +129,21 @@ internal static class GoToDefinitionHelpers
                 streamingPresenter.TryNavigateToOrPresentItemsAsync(
                     threadingContext, solution.Workspace, title, definitions, cancellationToken));
         }
+
+        public static bool TryGoToDefinition(
```
Comment (sharwell) [inline src/EditorFeatures/Core/GoToDefinition/GoToDefinitionHelpers.cs:133]: 💡 After this merges, can remove the previous `TryGoToDefinition` method as a separate follow-up. You have the option of either filing a follow-up issue or submitting a PR to just remove it in favor of the new method.

**Comment by vatsalyaagrawal** [comment]
@allisonchou headsup for 16.10.P2

**Comment by Cosifne** [comment]
This PR should be ready when https://github.com/dotnet/roslyn/pull/52379 is checked in

**Review by sharwell — APPROVED** [review]

**Pull request merged**